File: frontend/src/types/kudos.ts
/** Shared domain types for the Kudo frontend. */

export interface LeaderboardEntry {
  rank: number;
  userId: string;
  name: string;
  kudosCount: number;
}

/** A single kudo event as returned by the API. */
export interface Kudo {
  id: string;
  workspaceId?: string;
  fromUserId: string;
  fromUserName: string;
  fromAvatarUrl?: string;
  toUserId: string;
  toUserName: string;
  toAvatarUrl?: string;
  kudosCount: number;
  message: string;
  channel?: string;
  createdAt: string;
}

export type KudoItem = Kudo;

export interface KudosStats {
  receivedThisWeek: number;
  receivedThisMonth: number;
  givenThisWeek: number;
  givenThisMonth: number;
  streak?: number;
  quotaRemaining?: number;
}

/** Shape of GET /api/me/kudos */
export interface MyKudosResponse {
  received: Kudo[];
  given: Kudo[];
  stats: KudosStats;
}

export interface ChannelStat {
  name: string;
  count: number;
}

export interface TopUser {
  rank: number;
  userId: string;
  name: string;
  kudosCount: number;
}

export interface HomeData {
  stats: KudosStats;
  recentKudos: Kudo[];
  topUsers: TopUser[];
  channelStats: ChannelStat[];
}

export interface MyKudosData {
  received: Kudo[];
  given: Kudo[];
  stats: KudosStats;
}

export interface LeaderboardData {
  entries: LeaderboardEntry[];
  period: string;
}

/** Shape of GET|PUT /api/admin/settings */
export interface AdminSettings {
  emoji: string;
  currencySingular: string;
  currencyPlural: string;
  dailyAllowance: number;
}

/** Client-side validation errors for AdminSettings. */
export type AdminSettingsErrors = Partial<Record<keyof AdminSettings, string>>;

export type PeriodFilter = 'week' | 'month' | 'allTime';
export type LeaderboardTab = 'receivers' | 'givers';

File: backend/internal/kudos/kudos.go
package kudos

import (
	"errors"
	"time"
)

var ErrSelfKudo = errors.New("cannot give a kudo to yourself")
var ErrQuotaExceeded = errors.New("daily kudo quota exceeded")

type Kudo struct {
	ID           string    `json:"id"`
	WorkspaceID  string    `json:"workspaceId"`
	FromUserID   string    `json:"fromUserId"`
	FromUserName string    `json:"fromUserName"`
	ToUserID     string    `json:"toUserId"`
	ToUserName   string    `json:"toUserName"`
	Message      string    `json:"message"`
	Channel      string    `json:"channel"`
	EmojiCount   int       `json:"kudosCount"`
	CreatedAt    time.Time `json:"createdAt"`
}

type LeaderboardEntry struct {
	Rank        int    `json:"rank"`
	UserID      string `json:"userId"`
	UserName    string `json:"name"`
	WorkspaceID string `json:"workspaceId"`
	Total       int    `json:"kudosCount"`
}

// UserProfile is a lightweight user summary derived from kudo history.
type UserProfile struct {
	UserID   string `json:"userId"`
	UserName string `json:"name"`
}

type Stats struct {
	ReceivedThisWeek  int `json:"receivedThisWeek"`
	ReceivedThisMonth int `json:"receivedThisMonth"`
	GivenThisWeek     int `json:"givenThisWeek"`
	GivenThisMonth    int `json:"givenThisMonth"`
	Streak            int `json:"streak"`
	QuotaRemaining    int `json:"quotaRemaining"`
}

type ChannelStat struct {
	Name  string `json:"name"`
	Count int    `json:"count"`
}

type Service interface {
	Give(workspaceID, fromUserID, fromUserName, toUserID, toUserName, message, channel string, emojiCount int) (*Kudo, error)
	Leaderboard(workspaceID string, limit int, since *time.Time) ([]LeaderboardEntry, error)
	Received(workspaceID, userID string) ([]Kudo, error)
	Given(workspaceID, userID string) ([]Kudo, error)
}
